### PR TITLE
Levi Huff - Code Review #1 - IT3048C

### PR DESCRIPTION
**Pull request by lh1207**
# Analysis of the program.
NewsNow is a mobile application that provides users with the latest news and information from a variety of sources. It features a clean and user-friendly interface, making it easy to navigate and find relevant news. NewsNow aims to provide a convenient and efficient way for users to stay informed on the latest news and events happening around the world.

# Was the program available in UC Github on time?
Yes, all good

# Is the program documented/commented well enough for you to understand?
Yes! Their README gave me plenty of information needed to review the source code.

# Does the program compile?
After syncing build.gradle upon Android Studio launching, I was able to get it to compile.

# Rationale behind your changes.
- By linking the colors provided in the storyboard from colors.kt to theme.kt, we can ensure that the colors used in the app's theme are consistent with the colors used throughout the app. This follows the best Java coding practices of consistency and modularity. By separating the theme from the rest of the app's code, we can make it easier to maintain and update the app's design without having to search through the entire codebase. https://developer.android.com/develop/ui/views/theming/themes (documentation uses xml but I used the ui package instead). https://github.com/tron561/IT3048C-Repo/commit/011ba40124468baf641a2686892d15c23cdb1874 & https://github.com/tron561/IT3048C-Repo/commit/12c899c9e6101adb50e37a76a3f4baa7eacea4ba

- The date of an article is an important piece of information that users would want to see when browsing through news articles. Therefore, I added the publication date to the title and the description of the article in the app. This change applies to best Java coding practices as it improves the readability and usefulness of the app's content, which is a key aspect of user experience design. according to Google's Material Design Guidelines, "content should be clear and understandable" https://material.io/design. By providing the publication date in the article's title and description, we make it easier for users to quickly determine whether an article is relevant and current. https://github.com/tron561/IT3048C-Repo/commit/a2e617b350501b14085ade3e6febba7a60b32783

- The ArticleInfo composable now takes an Article object as its parameter instead of a String, which passed placeholder text. There is no need for placeholder text because there is an article that already exists per the Rhianna Super Bowl article. This change was made to follow the Java coding practice of using strong typing wherever possible. By using an Article object instead of a String, we can ensure that only valid Article objects are passed as parameters and avoid potential errors or unexpected behavior caused by passing invalid data types. https://developer.android.com/jetpack/compose/kotlin#dsl this article helps explain how to pass objects as parameters in a composable. https://github.com/tron561/IT3048C-Repo/commit/85d26ee2c80e7f59cc85cf9efe22084c5008e5ec

- The "full_description" var in the Article dto did not follow Kotlin naming conventions, mentioned here https://kotlinlang.org/docs/coding-conventions.html#naming-rules. Per the article, the var must use lowercaseUppercase conventions, ie: fullDescription https://github.com/tron561/IT3048C-Repo/commit/85d26ee2c80e7f59cc85cf9efe22084c5008e5ec

# Links to three specific commits that you made to your own group's GitHub repository before the end of sprint deadline.  
https://github.com/lh1207/LocoQuest/commit/1e57dc420f11afadc6cc7ff1ea4195b39186b052
https://github.com/lh1207/LocoQuest/commit/f54759f488db439d88fc3612cdc2bbf20ee583d4
https://github.com/lh1207/LocoQuest/commit/3b33ee6ebd9ef2a52f2e2415bea76436bf1c30ff

# Next, list three specific technical concepts that you learned from this code review.  Copy and paste samples to demonstrate your points.
I learned to only add necessary permissions in AndroidManifest. I almost added the foreground location permission, but then realized that it will not be utilized, therefore, not fitting the Definition of Done and unable to merge to master/main.
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools">
 
    <uses-permission android:name="android.permission.INTERNET"></uses-permission>
```

I learned that providing clear and concise information, even if it's a date to provide context, is best practice in code so that the user can understand the article and context without having to click on it first. This is shown in the return statement of the article dto
```kotlin
    override fun toString(): String {
        return "$title $fullDescription $pubDate"}
}
```

I learned that passing objects into a parameter is best practice because it avoids unexpected behaviors that invalid data types could cause, such as a String object that has placeholder text.
```kotlin
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            NewsNowTheme {
                // surface container background color
                Surface(color = MaterialTheme.colors.background,
                    modifier = Modifier.fillMaxWidth()){
                    ArticleInfo(
                        Article(
                            title = "Rihanna is BACK",
                            link = "https://example.com/rihanna",
                            creator = "Jane Doe",
                            full_description = "International superstar performs at the Super Bowl Halftime Show.",
                            pubDate = "2023-02-23",
                            id = 1
                        )
                    )
                }
            }
        }
    }
}
```

This example eliminates the need for placeholder text and puts the latest article in the MainActivity. There is no need for placeholder text because there is an article that already exists.

### CODE REVIEW

File: app/src/main/java/com/newsnow/MainActivity.kt [hunk 1]
```diff
@@ -7,37 +7,44 @@ import androidx.activity.compose.setContent
 import androidx.compose.foundation.layout.Column
 import androidx.compose.foundation.layout.fillMaxWidth
 import androidx.compose.material.MaterialTheme
-import androidx.compose.material.OutlinedTextField
 import androidx.compose.material.Surface
 import androidx.compose.material.Text
-import androidx.compose.runtime.*
+import androidx.compose.runtime.Composable
```
Comment (discospiff) [inline app/src/main/java/com/newsnow/MainActivity.kt:12]: Agreed.

Wildcard imports (*) are considered bad practice for a couple of reasons:

- Less documentation on what classes are being imported.
- If you have two wildcard imports, and they have a class with the same name, there will be a collision. Example: java.util.* and java.sql.* both have a class named Date. Different classes, same name. So, if you're using wildcards, which Date class should be used?
- It makes no difference in runtime performance, as the compiler will automatically optimize imports.

File: app/src/main/java/com/newsnow/MainActivity.kt [hunk 2]
```diff
@@ -46,53 +53,24 @@ fun ArticleInfo(name : String) {
         )
         Text(
             // title
-            text = "Rihanna is BACK",
+            text = article.title,
```
Comment (discospiff) [inline app/src/main/java/com/newsnow/MainActivity.kt:56]: If anything, I'd move the text to strings.xml

File: app/src/main/java/com/newsnow/dto/Article.kt [hunk 3]
```diff
@@ -9,12 +9,11 @@ data class Article(
     var title : String,
     var link : String,
     var creator : String,
-    var full_description : String,
+    var fullDescription : String,
```
Comment (discospiff) [inline app/src/main/java/com/newsnow/dto/Article.kt:12]: Good.

**Comment by discospiff** [comment]
Good suggestions.  I don't know that I'd merge the entire branch, but I do encourage the team to look at each change, and the corresponding comments, and decide what to harvest.

File: app/src/main/java/com/newsnow/dto/Article.kt (hunk 3 above)
Comment (francescasuba) [reply app/src/main/java/com/newsnow/dto/Article.kt:12]: We have implemented this suggestion already from a previous code review.

**Pull request closed without merging**